### ISSUE TITLE
No error when the return value of the function is an object

### ISSUE THREAD

**Issue by yamakeeeeeeeeen** [microsoft/TypeScript#41834]
**TypeScript Version:**  4.1.2

**Search Terms:** `function`, `return`

**Code**

```ts
const example: () => { name: string } = () => {
  return {
    name: 'jhon',
    age: '20',　　　　// No error 👎
  };
};
```

**Expected behavior:**

The 'age' is not included in the definition and therefore should get an error.

**Actual behavior:**

No errors.
You will get an error if you do the following.

```ts
const example = (): { name: string } => {
  return {
    name: 'jhon',
    age: '20',　　　　// error 👍
  };
};
```

**Comment by MartinJohns** [comment]
This is working as intended. Because you did not provide any type annotations for your arrow function the return type will be inferred. Then when assigning this function to your variable it will be checked for compatibility, which matches here.

**Comment by yamakeeeeeeeeen** [comment]
@MartinJohns 
Thank you for your answer.

The following code would result in an error on the editor and should show an error as well.
What do you think?

![image](https://user-images.githubusercontent.com/53547520/101283283-b614fa80-381c-11eb-8e47-917454b1b2cc.png)

**Comment by MartinJohns** [comment]
In your new example the types are not compatible. A `string` is not compatible with a `boolean`.

In your first example the types are compatible. `{ name: string; age: string }` is compatible with `{ name: string }`.

**Comment by yamakeeeeeeeeen** [comment]
@MartinJohns 
Understood.
Thank you for telling me.
